### PR TITLE
fix: Performance optimisation on flagging rules

### PR DESCRIPTION
**Pull request by andytaylorkainos**

### CODE REVIEW

File: workflow/flagging_rules_automation.py [hunk 1]
```diff
@@ -118,20 +118,40 @@ def apply_flagging_rule_to_all_open_cases(flagging_rule: FlaggingRule):
...
+
+            goods = GoodOnApplication.objects.filter(application_id__in=open_cases).values_list("good_id", flat=True)
+            flagging_rule.flag.goods.add(*goods)
 
         elif flagging_rule.level == FlagLevels.DESTINATION:
-            for case in open_cases:
-                apply_destination_flagging_rules_for_case(case, flagging_rule_queryset)
+            end_users = (
+                EndUserAdvisoryQuery.objects.filter(end_user__country_id=flagging_rule.matching_value)
+                .exclude(status__status__in=CaseStatusEnum.terminal_statuses() + [CaseStatusEnum.DRAFT])
```
Comment (BenAndrew-Kainos) [inline workflow/flagging_rules_automation.py]: Could you make CaseStatusEnum.terminal_statuses() + [CaseStatusEnum.DRAFT] a constant

**Comment by BenAndrew-Kainos** [comment]
Are the unit tests for this OK?

File: workflow/flagging_rules_automation.py [hunk 2]
```diff
@@ -118,20 +118,40 @@ def apply_flagging_rule_to_all_open_cases(flagging_rule: FlaggingRule):
...
+            open_cases = open_cases.filter(case_type__reference=flagging_rule.matching_value).values_list(
+                "id", flat=True
+            )
+
+            flagging_rule.flag.cases.add(*open_cases)
 
         elif flagging_rule.level == FlagLevels.GOOD:
-            for case in open_cases:
-                apply_good_flagging_rules_for_case(case, flagging_rule_queryset)
+            good_in_query = GoodsQuery.objects.exclude(
```
Comment (efahy) [inline workflow/flagging_rules_automation.py]: Please add comments throughout explaining the reason why you have to flag a bunch of different things, i.e. the different case types

Also please add a comment to the lines adding the flags to make it clear that you're not really modifying the flag, you're adding the flag to the good/case/destination

File: workflow/flagging_rules_automation.py (hunk 1 above)
Comment (andytaylorkainos) [reply workflow/flagging_rules_automation.py]: done

File: workflow/flagging_rules_automation.py (hunk 2 above)
Comment (efahy) [inline workflow/flagging_rules_automation.py]: Please add comments throughout explaining the reason why you have to flag a bunch of different things, i.e. the different case types

Also please add a comment to the lines adding the flags to make it clear that you're not really modifying the flag, you're adding the flag to the good/case/destination

**Pull request merged**